UAF in JSCreateScriptURLCallback via TrustedTypePolicy::~TrustedTypePolicy in WorkerOrWorkletThread::destroyWorkerGlobalScope

rniwa opened the pull request:
#### ff119714d9a57249197055fe7dbb5ba930fd220c
<pre>
UAF in JSCreateScriptURLCallback via TrustedTypePolicy::~TrustedTypePolicy in WorkerOrWorkletThread::destroyWorkerGlobalScope
<a href="https://bugs.webkit.org/show_bug.cgi?id=272193">https://bugs.webkit.org/show_bug.cgi?id=272193</a>
&lt;<a href="https://rdar.apple.com/122857425">rdar://122857425</a>&gt;

Reviewed by Chris Dumez.

Fix the bug that WorkerGlobalScope::prepareForDestruction doesn&apos;t remove WorkerGlobalScopeTrustedTypes
because it uses a different ASCIILiteral than the one used in the actual supplement.

* Source/WebCore/dom/WindowOrWorkerGlobalScopeTrustedTypes.cpp:
(WebCore::WorkerGlobalScopeTrustedTypes::supplementName):
(WebCore::WindowOrWorkerGlobalScopeTrustedTypes::workerGlobalSupplementName):
* Source/WebCore/dom/WindowOrWorkerGlobalScopeTrustedTypes.h:
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::prepareForDestruction):

Canonical link: <a href="https://commits.webkit.org/277095@main">https://commits.webkit.org/277095@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a80c92443bfa3be6a19b205ae666574b74696147

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46670 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25830 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49291 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49347 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42717 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48977 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30190 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23292 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47250 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22811 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/40207 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19312 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/20182 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41337 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 13 flakes 2 failures; Uploaded test results; 13 flakes 2 failures; Running compile-webkit-without-change") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4715 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42949 "Failed to checkout and rebase branch from PR 26874") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/41706 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51203 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21679 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/18062 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/45307 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/22967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/44260 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/23424 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6530 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22670 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez reviewed (approved):
LGTM

cdumez commented on Source/WebCore/workers/WorkerGlobalScope.cpp:
```diff
@@ -158,7 +159,7 @@ void WorkerGlobalScope::prepareForDestruction()
     WorkerOrWorkletGlobalScope::prepareForDestruction();
 
-    removeSupplement("WorkerGlobalScopeTrustedTypes"_s);
+    removeSupplement(WindowOrWorkerGlobalScopeTrustedTypes::workerGlobalSupplementName());
```
Ryosuke will perf A/B test a patch to make this less error prone. Hopefully, perf is OK and we can get rid of this class of bugs.

rniwa committed:
UAF in JSCreateScriptURLCallback via TrustedTypePolicy::~TrustedTypePolicy in WorkerOrWorkletThread::destroyWorkerGlobalScope
https://bugs.webkit.org/show_bug.cgi?id=272193
<rdar://122857425>

Reviewed by Chris Dumez.

Fix the bug that WorkerGlobalScope::prepareForDestruction doesn't remove WorkerGlobalScopeTrustedTypes
because it uses a different ASCIILiteral than the one used in the actual supplement.

* Source/WebCore/dom/WindowOrWorkerGlobalScopeTrustedTypes.cpp:
(WebCore::WorkerGlobalScopeTrustedTypes::supplementName):
(WebCore::WindowOrWorkerGlobalScopeTrustedTypes::workerGlobalSupplementName):
* Source/WebCore/dom/WindowOrWorkerGlobalScopeTrustedTypes.h:
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::prepareForDestruction):

Canonical link: https://commits.webkit.org/277095@main
